# Linear Regression - Exercise

The Programme for International Student Assessment (PISA) is a test given every three years to 15-year-old students from around the world to evaluate their performance in mathematics, reading, and science. This test provides a quantitative way to compare the performance of students from different parts of the world. In this exercise, we will predict the reading scores of students from the United States of America on the 2009 PISA exam.

The datasets 'pisa2009train.csv' and 'pisa2009test.csv' contain information about the demographics and schools for American students taking the exam, derived from 2009 PISA Public-Use Data Files distributed by the United States National Center for Education Statistics (NCES).

Each row in the datasets pisa2009train.csv and pisa2009test.csv represents one student taking the exam. The datasets have the following variables:

* grade: The grade in school of the student (most 15-year-olds in America are in 10th grade)

* male: Whether the student is male (1/0)

* raceeth: The race/ethnicity composite of the student

* preschool: Whether the student attended preschool (1/0)

* expectBachelors: Whether the student expects to obtain a bachelor's degree (1/0)

* motherHS: Whether the student's mother completed high school (1/0)

* motherBachelors: Whether the student's mother obtained a bachelor's degree (1/0)

* motherWork: Whether the student's mother has part-time or full-time work (1/0)

* fatherHS: Whether the student's father completed high school (1/0)

* fatherBachelors: Whether the student's father obtained a bachelor's degree (1/0)

* fatherWork: Whether the student's father has part-time or full-time work (1/0)

* selfBornUS: Whether the student was born in the United States of America (1/0)

* motherBornUS: Whether the student's mother was born in the United States of America (1/0)

* fatherBornUS: Whether the student's father was born in the United States of America (1/0)

* englishAtHome: Whether the student speaks English at home (1/0)

* computerForSchoolwork: Whether the student has access to a computer for schoolwork (1/0)

* read30MinsADay: Whether the student reads for pleasure for 30 minutes/day (1/0)

* minutesPerWeekEnglish: The number of minutes per week the student spend in English class

* studentsInEnglish: The number of students in this student's English class at school

* schoolHasLibrary: Whether this student's school has a library (1/0)

* publicSchool: Whether this student attends a public school (1/0)

* urban: Whether this student's school is in an urban area (1/0)

* schoolSize: The number of students in this student's school

* readingScore: The student's reading score, on a 1000-point scale

In [ ]:
import numpy as np
import pandas as pd

## Dataset size

Load the training and testing sets using the read.csv() function, and save them as variables with the names pisaTrain and pisaTest.

**How many students are there in the training set?**

In [ ]:
pisaTrain = pd.read_csv("pisa2009train.csv")
pisaTest = pd.read_csv("pisa2009test.csv")
pisaTrain.shape[0]

3663

## Summarizing the dataset

**Using pisaTrain, what is the average reading test score of males? or females?**

In [ ]:
pisaTrain.groupby('male')['readingScore'].mean()

male
0    512.940631
1    483.532479
Name: readingScore, dtype: float64

## Locating and removing missing values

**Which variables do not have any missing values in the training set?**

In [ ]:
pisaTrain.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3663 entries, 0 to 3662
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   grade                  3663 non-null   int64  
 1   male                   3663 non-null   int64  
 2   raceeth                3628 non-null   object 
 3   preschool              3607 non-null   float64
 4   expectBachelors        3601 non-null   float64
 5   motherHS               3566 non-null   float64
 6   motherBachelors        3266 non-null   float64
 7   motherWork             3570 non-null   float64
 8   fatherHS               3418 non-null   float64
 9   fatherBachelors        3094 non-null   float64
 10  fatherWork             3430 non-null   float64
 11  selfBornUS             3594 non-null   float64
 12  motherBornUS           3592 non-null   float64
 13  fatherBornUS           3550 non-null   float64
 14  englishAtHome          3592 non-null   float64
 15  comp

In [ ]:
pisaTrain.describe()

grade         male    preschool  expectBachelors     motherHS  \
count  3663.000000  3663.000000  3607.000000      3601.000000  3566.000000   
mean     10.089817     0.511057     0.722761         0.785893     0.879978   
std       0.554375     0.499946     0.447697         0.410259     0.325033   
min       8.000000     0.000000     0.000000         0.000000     0.000000   
25%      10.000000     0.000000     0.000000         1.000000     1.000000   
50%      10.000000     1.000000     1.000000         1.000000     1.000000   
75%      10.000000     1.000000     1.000000         1.000000     1.000000   
max      12.000000     1.000000     1.000000         1.000000     1.000000   

       motherBachelors   motherWork     fatherHS  fatherBachelors  \
count      3266.000000  3570.000000  3418.000000      3094.000000   
mean          0.348132     0.734454     0.859274         0.331933   
std           0.476451     0.441685     0.347789         0.470983   
min           0.000000     0.000000     0.000000         0.000000   
25%           0.000000     0.000000     1.000000         0.000000   
50%           0.000000     1.000000     1.000000         0.000000   
75%           1.000000     1.000000     1.000000         1.000000   
max           1.000000     1.000000     1.000000         1.000000   

        fatherWork  ...  englishAtHome  computerForSchoolwork  read30MinsADay  \
count  3430.000000  ...    3592.000000            3598.000000     3629.000000   
mean      0.853061  ...       0.871659               0.899389        0.289887   
std       0.354096  ...       0.334515               0.300855        0.453772   
min       0.000000  ...       0.000000               0.000000        0.000000   
25%       1.000000  ...       1.000000               1.000000        0.000000   
50%       1.000000  ...       1.000000               1.000000        0.000000   
75%       1.000000  ...       1.000000               1.000000        1.000000   
max       1.000000  ...       1.000000               1.000000        1.000000   

       minutesPerWeekEnglish  studentsInEnglish  schoolHasLibrary  \
count            3477.000000        3414.000000       3520.000000   
mean              266.208225          24.499414          0.967614   
std               148.403525           7.184348          0.177049   
min                 0.000000           1.000000          0.000000   
25%               225.000000          20.000000          1.000000   
50%               250.000000          25.000000          1.000000   
75%               300.000000          30.000000          1.000000   
max              2400.000000          75.000000          1.000000   

       publicSchool        urban   schoolSize  readingScore  
count   3663.000000  3663.000000  3501.000000   3663.000000  
mean       0.933934     0.384930  1369.316767    497.911403  
std        0.248431     0.486645   869.983618     95.515153  
min        0.000000     0.000000   100.000000    168.550000  
25%        1.000000     0.000000   712.000000    431.705000  
50%        1.000000     0.000000  1212.000000    499.660000  
75%        1.000000     1.000000  1900.000000    566.230000  
max        1.000000     1.000000  6694.000000    746.000000  

[8 rows x 23 columns]

grade, male, publicSchool, urban, and readingScore do not have any missing value. 

Linear regression discards observations with missing data, so we will remove all such observations from the training and testing sets. **How many observations are now in the training and test sets?**

In [ ]:
pisaTrain=pisaTrain.dropna(axis=0,how='any') #drop all rows that have any NaN values
pisaTrain.shape[0]

2414

In [ ]:
pisaTest=pisaTest.dropna(axis=0,how='any') #drop all rows that have any NaN values
pisaTest.shape[0]

990

## Handling categorical variables

Categorical variables are variables that take on a discrete set of values. 'Color' variable with levels of 'red', 'green', and 'blue', is an example. To include a categorical variable in a linear regression model, we define one level as the "reference level" and add a binary variable for each of the remaining levels. In this way, a categorical variable with n levels is replaced by n-1 binary variables. 

As an example, consider the 'color' variable. If "green" were the reference level, then we would add binary variables "colorred" and "colorblue" to a linear regression problem. All red examples would have colorred=1 and colorblue=0. All blue examples would have colorred=0 and colorblue=1. All green examples would have colorred=0 and colorblue=0.

Now, consider the variable "raceeth" in our problem, which has levels "American Indian/Alaska Native", "Asian", "Black", "Hispanic", "More than one race", "Native Hawaiian/Other Pacific Islander", and "White". Because it is the most common in our population, we will select White as the reference level. **How many binary variables should we include in a linear regression model?**

In [ ]:
pisaTrain['raceeth'].value_counts()

White                                     1470
Hispanic                                   500
Black                                      228
Asian                                       95
More than one race                          81
Native Hawaiian/Other Pacific Islander      20
American Indian/Alaska Native               20
Name: raceeth, dtype: int64

There are 7 different values in the variable. So we should include 6 binary variables. 

In [ ]:
#creat dummy variables and set 'White' to be the reference level
pisaTrain=pd.get_dummies(pisaTrain, prefix=['raceeth'], columns=['raceeth'])
pisaTrain=pisaTrain.drop(columns='raceeth_White') # or pisaTrain=pisaTrain.drop('raceeth_White', axis=1)

## Build a model

Now, build a linear regression model (call it lmScore) using the training set to predict readingScore using all the remaining variables.

In [ ]:
from sklearn import linear_model
from sklearn import metrics

In [ ]:
#linear regression
X_train = pisaTrain.drop('readingScore', axis=1) 
y_train = pisaTrain['readingScore']
lmScore = linear_model.LinearRegression()
lmScore.fit(X_train, y_train)

LinearRegression()

**What is the RMSE of lmScore on the training set? (Poll) How good do you think the current model is?**

In [ ]:
y_train_predictions=lmScore.predict(X_train)

#What are MSE and RMSE of lmScore on the training set?
print('MSE:', metrics.mean_squared_error(y_train, y_train_predictions))
print('RMSE:', np.sqrt(metrics.mean_squared_error(y_train, y_train_predictions)))

MSE: 5382.5041370658955
RMSE: 73.3655514329845


RMSE is 73. The average score is 498. Thus, RMSE is 14.7% of the average score. It seems a bit large. So we should do more to improve the prediction performance. 

## Comparing predictions for similar students

Consider two students A and B. They have all variable values the same, except that student A is in grade 11 and student B is in grade 9. **What is the predicted reading score of student A minus the predicted reading score of student B? (Poll)**

In [ ]:
# coefficients

coeff_df = pd.DataFrame(lmScore.coef_,X_train.columns,columns=['Coefficient'])
coeff_df

Coefficient
grade                                             29.542707
male                                             -14.521653
preschool                                         -4.463670
expectBachelors                                   55.267080
motherHS                                           6.058774
motherBachelors                                   12.638068
motherWork                                        -2.809101
fatherHS                                           4.018214
fatherBachelors                                   16.929755
fatherWork                                         5.842798
selfBornUS                                        -3.806278
motherBornUS                                      -8.798153
fatherBornUS                                       4.306994
englishAtHome                                      8.035685
computerForSchoolwork                             22.500232
read30MinsADay                                    34.871924
minutesPerWeekEnglish                              0.012788
studentsInEnglish                                 -0.286631
schoolHasLibrary                                  12.215085
publicSchool                                     -16.857475
urban                                             -0.110132
schoolSize                                         0.006540
raceeth_American Indian/Alaska Native            -67.277327
raceeth_Asian                                     -4.110325
raceeth_Black                                    -67.012347
raceeth_Hispanic                                 -38.975486
raceeth_More than one race                       -16.922522
raceeth_Native Hawaiian/Other Pacific Islander    -5.101601

The predicted reading score of student A minus the predicted reading score of student B is 29.5427070917684*2=59.09

## Interpreting coefficients

What is the meaning of the coefficient associated with variable raceeth_Asian? 

1. Predicted average reading score of an Asian student

2. Difference between the average reading score of an Asian student and the average reading score of a white student

3. Difference between the average reading score of an Asian student and the average reading score of all the students in the dataset

4. Predicted difference in the reading score between an Asian student and a white student who is otherwise identical

## Predicting on unseen data

Use the lmScore model to predict the reading scores of students in pisaTest. 

What is the range between the maximum and minimum predicted reading score on the test set?

In [ ]:
#creat dummy variables and set 'White' to be the reference level for pisaTest
pisaTest=pd.get_dummies(pisaTest, prefix=['raceeth'], columns=['raceeth'])
pisaTest=pisaTest.drop(columns='raceeth_White')
#predict
X_test = pisaTest.drop('readingScore', axis=1) 
y_test_predictions = lmScore.predict(X_test)
print(max(y_test_predictions)-min(y_test_predictions))

284.46831179513725


## Test set MSE and RMSE

What is the RMSE of lmScore on the test set? 

In [ ]:
y_test=pisaTest['readingScore']

#What are MSE and RMSE of lmScore on the test set?
print('MSE:', metrics.mean_squared_error(y_test, y_test_predictions))
print('RMSE:', np.sqrt(metrics.mean_squared_error(y_test, y_test_predictions)))

MSE: 5820.285223378207
RMSE: 76.29079383109215


## How can we improve the model performance on the test set?

There are several things we can do. We can do different preprocessing tasks like different variable transformation. We can also try to create new variables like interactions and use them in a regression model. We can also try different machine learning models. We can do a hyperparameter tuning.  